src/Futhark/Optimise/ArrayShortCircuiting/ArrayCoalescing.hs

```diff
@@ -1326,30 +1338,120 @@ mkCoalsHelper3PatternMatch pat e lutab td_env successCoals_tab activeCoals_tab i
                               opts'
                               mempty
                        in M.insert m_b coal_etry acc
-            _ -> acc
-mkCoalsHelper3PatternMatch _ _ _ _ _ _ _ =
-  error "In ArrayCoalescing.hs, fun mkCoalsHelper3PatternMatch: Unreachable!!!"
+            else acc
+
+-- | Information about a particular short-circuit point
+type SSPointInfo =
+  ( CoalescedKind,
+    IxFun -> IxFun,
+    VName,
+    VName,
+    IxFun,
+    VName,
+    VName,
+    IxFun,
+    PrimType,
+    Shape
+  )
+
+-- | Given an 'Op', return a list of potential short-circuit points
+genSSPointInfoSeqMem ::
+  LUTabFun ->
+  TopdownEnv SeqMem ->
+  ScopeTab SeqMem ->
+  Pat (VarAliases, LetDecMem) ->
+  Op (Aliases SeqMem) ->
+  Maybe [SSPointInfo]
+genSSPointInfoSeqMem _ _ _ _ _ =
+  Nothing
+
+-- | Given an 'Op', return a list of potential short-circuit points
+--
+-- For 'SegOp', we currently only handle 'SegMap', and only under the following
+-- circumstances:
+--
+--  1. The 'SegMap' has only one return/pattern value.
```

Comment:
How difficult would this be to improve?  Doesn't have to be part of this PR, of course.

Reply:
We would have to attempt to match candidate sources to the patterns in the destination. We would also need to pass in the `InhibitTab` from `BotUpEnv` in order to keep track of which combinations have failed so far. I don't think it would be that hard to do.